### ISSUE TITLE
Enables `implements` on abstract classes with protected members (usefull for generic mixins)

### ISSUE THREAD

**Issue by denis-migdal** [microsoft/TypeScript#61161]
### 🔍 Search Terms

implement abstract class with protected

### ✅ Viability Checklist

- [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [x] This wouldn't change the runtime behavior of existing JavaScript code
- [x] This could be implemented without emitting different JS based on the types of the expressions
- [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [x] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [x] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

Enable doing the following (currently this raises an error):

```ts
abstract class Base {
  protected abstract foo(): void
}

class B implements Base {
  protected foo() {}
}
```

### 📃 Motivating Example

This would be useful to declare "protected" interfaces, when we can't use `extends` as the child class is already extending another.

```ts
// interface
abstract class Base {
  protected abstract foo(): void
}

class C {}

class B extends C implements Base {
  protected foo() {}
}
```

Such need is frequent when using generic mixins: indeed, type manipulations inside generic functions (and mixins) are often quite troublesome, with unwanted type expansions/generalization (?). Then this is often better explicitly describe the returning type.

To do so, `abstract class` is the only way to properly describe the properties added by the mixin. However, if such class has protected properties, this raises TS error as classes with protected properties are no longer duck typing but "nominal" (?).

```ts
type Cstr<T> = new(...args:any[]) => T;
type Merge<A extends abstract new(...args:any[])=>any,
           B extends new(...args:any[])=>any
          > = Omit<A & B, "new"> & (new() => (InstanceType<A> & InstanceType<B>))

function foo<T extends Cstr<C>>(klass: T): Merge<typeof Base, T> {
  return class B extends klass implements Base {
    protected foo() {}
  }
}

class X extends foo(C) {}
```

### 💻 Use Cases

1. What do you want to use this for?

Easily describing generic mixins return type.

2. What shortcomings exist with current approaches?

Can't describe protected properties.

3. What workarounds are you using in the meantime?

Dirty casting, removing some type checks inside the generic mixin.

**Comment by denis-migdal** [comment]
Another related issue:
```ts
abstract class B {
    protected abstract foo(): void;
}

function foo(b: B) {
    // @ts-ignore
    b.foo(); // use non-public API.
    // we can do better for type safety, but for the sake of the example let's do it that way.
}

class A { // could also "implements B"
    protected foo() {}
}
    
foo( new A() ); // error
```

> Argument of type 'A' is not assignable to parameter of type 'B'.
> Property 'foo' is protected but type 'A' is not a class derived from 'B'.(2345)

`foo()` shouldn't care if `A` is really a `B` or not.

I don't even know why this check is here in the first place.

_Note:_ I was this close to open a new issue...
Additional search terms (to improve issue search by keywords):
- Property is protected but type is not a class derived from (2345) 
- Property is protected but type is not a class derived from (2720) 
- Property protected not a class derived from
- implements abstract class with protected members
